### PR TITLE
ci(test-cli): migrate to gha workflow

### PR DESCRIPTION
**Pull request by loksonarius**
## Description

Migrating test-cli CircleCI workflow into an independent GHA workflow as
part of general migration out of CircleCI. The steps are pretty much the
same with the following differences from the current CircleCI workflow:

- test-cli will now run all tests across both Windows and Ubuntu for
  each Python version
- the test environment is entirely setup with just Python and pip -- no
  conda installation of anything

## Test Plan

Run workflow in both CircleCI and GHA with a PR and double-check output for differences.

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by loksonarius** [comment]
So comparing the output of the two runs for Windows@Py3.7, it seems the CircleCI workflow is hiding some warnings that show up in GHA -- the same amount of tests still pass and skip though.
Ref Circle: https://app.circleci.com/pipelines/github/determined-ai/determined/24556/workflows/1785af75-9a15-4075-9799-0c85f03b4d6a/jobs/795328
Ref GHA: https://github.com/determined-ai/determined/runs/7400922371?check_suite_focus=true

Output for the Ubuntu runners was the same, and run time was comparable across both platforms (sub-minute at worst).
Windows run times were comparable as well within a minute of each other for Python 3.7.

File: .github/workflows/test-cli.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Install det dependencies
+        run: |
+          python --version
+          python -m pip install --upgrade pip
+          pip --version
+          pip install wheel cryptography patch-ng
+          cd harness; python setup.py bdist_wheel -d ../build
+      - name: Install determined
+        run: |
+          pip install --find-links build determined==0.18.5.dev0
```
Comment (loksonarius) [inline .github/workflows/test-cli.yml:37]: Working right now on figuring out the source of truth for this value -- if anyone knows, please comment 🙏

Comment (rb-determined-ai) [reply .github/workflows/test-cli.yml:37]: you mean the version?  We maintain it with bumpversion, see `.bumpversion.cfg` I think.

Comment (loksonarius) [reply .github/workflows/test-cli.yml:37]: So I should just use the value of `current_version` there?

File: .github/workflows/test-cli.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Setup Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install det dependencies
+        run: |
+          python --version
+          python -m pip install --upgrade pip
+          pip --version
+          pip install wheel cryptography patch-ng
```
Comment (rb-determined-ai) [inline .github/workflows/test-cli.yml:33]: there used to be context as to why these two lines were here... maybe keep the comment that explained them?

If you want to install the extra deps on linux too, that's fine.  But leave the explanations please.

File: .github/workflows/test-cli.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          cd harness; python setup.py bdist_wheel -d ../build
+      - name: Install determined
+        run: |
+          pip install --find-links build determined==0.18.5.dev0
+          det --help
```
Comment (rb-determined-ai) [inline .github/workflows/test-cli.yml:38]: preserve comment please

File: .github/workflows/test-cli.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          pip freeze --all
+          pip check || true
```
Comment (rb-determined-ai) [inline .github/workflows/test-cli.yml:40]: ditto

File: .github/workflows/test-cli.yml (hunk 1 above)
Comment (rb-determined-ai) [reply .github/workflows/test-cli.yml:37]: I mean, I think it you configure bumpversion to update this file, then it should be safe to have the literal `0.18.5.dev0` written into the file.

Comment (loksonarius) [reply .github/workflows/test-cli.yml:37]: Ah, that's an idea as well. While I don't like the idea of potentially adding noise when version bumping, it seems much more reasonable to make the relationship between the files explicit by adding it to that config file. Thanks for the suggestion!

Comment (dannysauer) [reply .github/workflows/test-cli.yml:37]: (quietly mentions that having a versioned version file is kinda silly when the repo metadata already tracks versions)

**Review by rb-determined-ai — APPROVED** [review]

**Review by dannysauer — APPROVED** [review]
I tend to think it's the base image maintainer's problem to keep pip up to date, and not something we should spend action cycles on (re: `pip install --upgrade pip`).  But that's just something to consider for future updates; I'm assuming this is a near-direct copy of existing code for the first pass.  So we can leave it.  This LGTM otherwise.  Thanks!

**Comment by rb-determined-ai** [comment]
> I tend to think it's the base image maintainer's problem to keep pip up to date, and not something we should spend action cycles on (re: pip install --upgrade pip)

I would agree with you in principle, but maybe the only useful coverage these tests have ever given us is early warning when upstream dependencies have broken.

Upgrading pip itself hasn't empirically proven to be useful, but in the same vein as always updating the latest python dependencies every test run, I think this is a decent solution.

I'll admit that there's no particular reason why the dependency install testing needs to be part of test-cli, and it wouldn't hurt my feelings to move it to a dedicated test, and leave the pip installing in the ci base image.

**Pull request merged**